python-ecosys/urequests/urequests.py

```diff
@@ -110,7 +110,7 @@ def request(
             if chunked_data:
                 s.write(b"Transfer-Encoding: chunked\r\n")
             else:
-                s.write(b"Content-Length: %d\r\n" % len(data))
+                s.write(b"Content-Length: %d\r\n" % len(data.encode()))
```

Comment:
This looks like a tiny change but what it means is that we need to unconditionally make a complete copy of the payload in RAM, only to then compute its length. This will significantly impact the memory overhead of using urequests for posting data.

I note that in the upstream "requests" libraries, it is invalid to pass a string (not bytes) as the `data` argument if that string contains non-`Latin-1` characters.

e.g.

```py
>>> requests.get('http://localhost:8000/', data='⛵')
...
UnicodeEncodeError: 'latin-1' codec can't encode character '\u26f5' in position 0: Body ('⛵') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
```

I'm not sure if MicroPython has an easy way to detect this case though (without doing something like `len(foo.encode()) != len(foo)` which obviously has the same problem.

But it's definitely not ideal what it does at the moment where it sends the incorrect Content-Length header....

Reply:
This also affects json payloads with multibyte utf-8 characters. The json gets truncated when sent, because the length is off.

Does encoding a string really allocate new data? I thought encoding just casts the raw string data as bytes?

See also issue #784 for aiohttp in micropython-lib